MaxText/inference/page_manager.py

```diff
@@ -1,232 +1,582 @@
+    page_map: A `jnp.ndarray` of shape `[max_page_groups, max_pages_per_group]`.
+      This array maps each page group to the indices (within the global pool)
+      of its allocated pages. Entries beyond `num_pages_used` for a group are invalid.
+    num_pages_used: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      tracks the number of pages currently allocated to each page group. This
+      determines the valid entries in `page_map` for each group.
+    sequence_lengths: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      stores the current true length of each sequence (in tokens) associated
+      with a page group.
+    active_page: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      stores the global index of the *currently active* page (the page where the
+      next token will be written) for each page group. Only valid if the
+      corresponding `has_active_page` is True.
+    has_active_page: A `jnp.ndarray` of shape `[max_page_groups]`. Boolean mask
+      indicating whether a page group currently represents an active sequence
+      and thus whether its `active_page` and `active_page_position` entries
+      are meaningful.
+    active_page_position: A `jnp.ndarray` of shape `[max_page_groups]`. This array
+      stores the index (offset, 0 to tokens_per_page-1) of the next available
+      token *within the `active_page`* for each page group. Only valid if
+      `has_active_page` is True.
   """
 
-  page_status: Array
-  page_map: Array
-  sequence_lengths: Array
-  num_pages_used: Array
-  current_page: Array
-  current_page_position: Array
+  page_status: PagesInt1d
+  page_map: GroupsPagesInt2d
+  num_pages_used: GroupsInt1d
+  sequence_lengths: GroupsInt1d
+  active_page: GroupsInt1d
+  has_active_page: GroupsBool1d
+  active_page_position: GroupsInt1d
 
 
 def initialize_page_state(
     num_pages: int,
-    num_slots: int,
-    max_pages_per_slot: int,
+    max_page_groups: int,
+    max_pages_per_group: int,
 ) -> PageState:
+  """Creates and initializes a global `PageState` object.
+
+  All pages in the global pool are initially marked as free (status 0), except
+  for page 0 which is marked used as a workaround. No pages are assigned to any
+  page group. Sequence lengths and page usage counts are initialized to zero.
+  Active page tracking is also reset.
+
+  Args:
+    num_pages: The total number of available pages in the global pool.
+    max_page_groups: The maximum number of page groups (concurrent sequences/requests)
+      the system can track.
+    max_pages_per_group: The maximum number of pages that can be allocated to
+      a single page group (determines the size of the second dimension of `page_map`).
+
+  Returns:
+    An initialized `PageState` object with all values set to their defaults (zeros/False).
+  """
+  initial_page_status = jnp.zeros((num_pages,), dtype=jnp.int32)
+  initial_page_status = initial_page_status.at[0].set(1)  # Workaround page 0
```

Comment:
Maybe add a TODO for any issue with using page 0, so we can follow up and have it fixed.

Reply:
Will add.